static_frame/core/index_hierarchy_set_utils.py

```diff
@@ -0,0 +1,480 @@
+import typing as tp
+from functools import partial
+
+import numpy as np
+from arraykit import mloc
+
+from static_frame.core.container_util import index_many_to_one
+from static_frame.core.exception import ErrorInitIndex
+from static_frame.core.index import Index
+from static_frame.core.index_hierarchy import IndexHierarchy
+from static_frame.core.loc_map import HierarchicalLocMap
+from static_frame.core.util import DtypeSpecifier
+from static_frame.core.util import IndexConstructor
+from static_frame.core.util import ManyToOneType
+from static_frame.core.util import intersect1d
+from static_frame.core.util import setdiff1d
+from static_frame.core.util import ufunc_unique1d
+from static_frame.core.util import ufunc_unique1d_indexer
+
+
+class ValidationResult(tp.NamedTuple):
+    indices: tp.List[IndexHierarchy]
+    depth: int
+    any_dropped: bool
+    any_shallow_copies: bool
+    name: tp.Hashable
+    index_constructors: tp.List[IndexConstructor]
+
+
+def _get_shallow_copy_key(ih: IndexHierarchy) -> tp.Hashable:
+    # calling id(...) on numpy arrays is not always reliable, so instead,
+    # we will use the underlying array properties to determine the ID for this
+    # index hierarchy, with the desire to encounter duplicate keys for shallow copies
+
+    result: tp.List[tp.Union[int, tp.Tuple[int, ...]]] = []
+
+    for block in ih._blocks._blocks:
+        # You need all three of these to uniquely identify a numpy array
+        result.append(mloc(block))
```

Comment:
I am actually fairly confident that this is correct.

The reason why, is because `array.base` only lets you know that you have a memory view, but it doesn't tell anything about the view itself.

This approach looks at `mloc`, which will find if we share same memory location, and strides/size will ensure that the view is the same.

Additionally, @michaeljpeters found that calling `id(np.ndarray)` multiple times can actually change what value is returned, so I really wanted to avoid calling `id(...)`.

Reply:
Agreed. Just see my question regarding if we need to include dtype in the comparison.